### PR TITLE
[inductor] correctly handle resize for AOTInductor wrapper calls

### PR DESCRIPTION
**Pull request by chenyang78**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107848

When generating a wrapper call, we may have implicit resize applied to
the kernel's output. For example, for addmm(3d_tensor, 2d_tensor),
its output buffer is resized to a 2d tensor. This triggers a warning from
Aten's resize_output op:

    "UserWarning: An output with one or more elements was resized since it had...
    This behavior is deprecated, and in a future PyTorch release outputs will
    not be resized unless they have zero elements..."

More importantly, the output shape is not the same as we would expect, i.e.
2d tensor v.s. 3d tensor.

This PR fixed the issue by injecting resize_(0) before calling the relevant
kernel and resize_(expected_shape) after the kernel call.

We also fixed a minor typo in the PR.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by desertfire — COMMENTED** [review]
`resize` feels hacky here. Can we make the output shape query return the correct shape in the first place?

**Comment by chenyang78** [comment]
> `resize` feels hacky here. Can we make the output shape query return the correct shape in the first place?

The output shape query returns the correct shape. It's the intermediate wrapper call that requires a different shape than the "correct" one. Also, the comment at (https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Resize.cpp#L29-L35) seems to indicate using ```resize``` is the recommended way?

**Comment by desertfire** [comment]
But your changes only affect the case where the tensor is a graph output. Maybe all you need is to make sure that ReinterpretView is codegen-ed as resize_ on the corresponding output tensor.  Cross reference https://github.com/pytorch/pytorch/pull/106564/files

**Comment by chenyang78** [comment]
> But your changes only affect the case where the tensor is a graph output. Maybe all you need is to make sure that ReinterpretView is codegen-ed as resize_ on the corresponding output tensor. Cross reference https://github.com/pytorch/pytorch/pull/106564/files

You meant that instead of generating something like below: 

```
auto buf0 = outputs[0]; buf0.resize_(0);
at::addmm_out(buf0, arg1_1, at::as_strided(arg2_1, {16L, 16L}, {16L, 1L}, 0L), at::as_strided(arg0_1, {16L, 6L}, {1L, 16L}, 0L), 1, 1);
...
outputs[0].resize_({2, 8, 6});
```

we generate the following form:

```
auto buf0 = outputs[0]; buf0.resize_({16, 6});
at::addmm_out(buf0, arg1_1, at::as_strided(arg2_1, {16L, 16L}, {16L, 1L}, 0L), at::as_strided(arg0_1, {16L, 6L}, {1L, 16L}, 0L), 1, 1);
...
outputs[0].resize_({2, 8, 6});
```

We could do this, although I think both forms achieve the same effect?

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -1094,14 +1095,19 @@ def generate_return(self, output_refs):
...
+                if hasattr(output, "get_name"):
+                    name = output.get_name()
+                    if name in self.outputs_need_copy:
+                        output_as_strided = output.codegen_reference()
+                        self.wrapper_call.writeline(
+                            f"outputs[{idx}].copy_({output_as_strided});"
+                        )
+                    resize_to = self.resized_outputs.get(name, None)
+                    if resize_to is not None:
+                        resize_to_args = ", ".join(pexpr(d) for d in resize_to)
```
Comment (desertfire) [inline torch/_inductor/codegen/wrapper.py:1107]: Nit: it mostly won't make a difference but I think you want `cexpr` here.

**Review by desertfire — APPROVED** [review]

**Review by jansel — APPROVED** [review]
Should passing wrong sized outputs be an error?  It can't be good for perf to do this...

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (chenyang78) [reply torch/_inductor/codegen/wrapper.py]: Good point. Fixed. Thanks!

**Comment by chenyang78** [comment]
> Should passing wrong sized outputs be an error? It can't be good for perf to do this...

I think it depends on how the user passes the output. In some cases, it seems to be more flexible. For example, if the output tensor has any dynamic dimension, the user may preallocate a tensor with an appropriate maximum capacity. In such a case, we need to resize the tensor to its right form at runtime.

**Comment by chenyang78** [comment]
> I was worrying `buf0.resize_(0)` would have any impact on the underlying storage, but reading its implementation suggests it's ok.

I choose to use the form of ```resize_({16, 6})``` instead of ```resize_(0)```, because I think the explicit form is slightly better.

**Pull request closed without merging**